### DISCUSSION TITLE
LineValueSampler: VectorPostProcessor in PostProcessor

### DISCUSSION THREAD

**Discussion by TFdvd** [idaholab/moose#28827] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hello MOOSE community! 
I'm intrested in the behaviour of LineValueSampler. In particular, the latter is a VectorPostProcessor and I get a .csv for each iteration. Is it possible to pass LineValueSampler from a VectorPostProcessor to a PostProcessor so that I can manipulate its behaviour to obtain a single .csv? Specifically, LineValueSampler in VectorPostProcessor returns me a .csv with a number of results = num_points (requested parameter) . My idea is to save this dataset, but to manipulate it in the PostProcessor area to get a single value of my interest. 
Do you have any suggestions for following this path? Maybe this is a silly question, but I don't have much experience. 
Thanks for your time and patience :)

**Comment by GiudGiud** [comment]
Hello

All the postprocessors get lumped into the same CSV file by default.
However you can create a cutom CSV output and use the 'show' parameter to select which postprocessors to print there.

You can retrieve the vector values of of a VPP in a custom postprocessor using 
```
getVectorPostprocessorValue("name of VPP", "which vector")
```
this return a reference, you should save it in an attribute of your custom PP

**Reply by TFdvd** [reply]
Hello @GiudGiud , thanks for your answer! I got your suggestion and I implemented it into a .h and .c that I will attach below. 
From LineValueSampler I obtained a .csv with [variable][id][x][y][z] as output. 
In the .h and .c below, my purpose is to save [variable] and possibly [y], in a Postprocessor in order to compare [variable] in this way: 

```
(std::abs(_c_var[i] - _threshold) < _tolerance)`
```
I'm not sure about the result because in the terminal got these error and notes: 
```
error: invalid new-expression of abstract class type 'NeckValue'
  962 |     { return unique_ptr<_Tp>(new _Tp(std::forward<_Args>(__args)...)); }
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


note:   because the following virtual functions are pure within 'NeckValue':
   29 | class NeckValue : public GeneralPostprocessor
      |       ^~~~~~~~~

note:     'virtual PostprocessorValue Postprocessor::getValue() const'
   39 |   virtual PostprocessorValue getValue() const = 0;
      |                              ^~~~~~~~
```
So I'm not able to proceed. Do you have some suggestion to correct and continue my custom Postprocessor? 

```
`NeckValue.h`

`#ifndef NECKVALUE_H
#define NECKVALUE_H

#include "GeneralPostprocessor.h"


class NeckValue; 


class NeckValue : public GeneralPostprocessor
{
public:
  
  Real computeValue();

  static InputParameters validParams();
  NeckValue(const InputParameters & parameters);
  
  virtual void initialize() override;
  virtual void execute() override;



  

protected:

  
  std::vector<Real> _c_var;
  const Real _threshold;
  const Real _tolerance;
 


};

#endif //NECKVALUE_H`
```

----------------------------------

```
`NeckValue.C`


`#include "NeckValue.h"
#include "NeckLineValueSampler.h"  
#include "SystemBase.h" 
using namespace std;    

registerMooseObject("ParrotApp", NeckValue);

InputParameters
NeckValue::validParams()
{
  InputParameters params = GeneralPostprocessor::validParams();

params.addRequiredParam<Real>("threshold", "Seconda variabile");
params.addRequiredParam<Real>("tolerance", "Terza variabile");

params.addRequiredParam<VectorPostprocessorName>("vpp", "The name of the VectorPostprocessor to pull the data from."); //new
params.addRequiredParam<std::string>("vector_name","The name of the vector to use from the VectorPostprocessor"); //new
params.addRequiredParam<std::vector<Real>>("c_var", "Lista di valori per c_var");
  return params;
}


NeckValue::NeckValue(
  const InputParameters &parameters)  

    : GeneralPostprocessor(parameters), 
    _threshold(getParam<Real>("threshold")),
    _tolerance(getParam<Real>("tolerance")),
    _c_var(getParam<std::vector<Real>>("c_var"))
{}

void 
NeckValue::initialize()
{

}
void 
NeckValue::execute()
{
  
}


  Real NeckValue::computeValue() { 
  Real value = 0.0; 
 
  
  VectorPostprocessorValue c_var = getVectorPostprocessorValueByName(getParam<VectorPostprocessorName>("vpp"),getParam<std::string>("vector_name"));
  for (unsigned int i = 0; i < 100; i++)
  {
    if (std::abs(_c_var[i] - _threshold) < _tolerance) 
    {
      value = c_var[i];
    }
  }
  return value;
}

`
```

Thanks @GiudGiud for your time and patience!

**Reply by GiudGiud** [reply]
hello

this is the message to look for
```
note: because the following virtual functions are pure within 'NeckValue':
29 | class NeckValue : public GeneralPostprocessor
| ^~~~~~~~~
note: 'virtual PostprocessorValue Postprocessor::getValue() const'
39 | virtual PostprocessorValue getValue() const = 0;
```

bascially you need to implement getValue
so in the header add
```
PostprocessorValue getValue() const override;
```
and in the source add
```
PostprocessorValue getValue() const
{
return _value;
}
```

in computeValue(), you should save `value` to a `_value` attribute

**Reply by TFdvd** [reply]
Hello @GiudGiud, thanks for your help! I need again support as regarding a doubt. I have a Postprocessor in which I manipulate a vector obtained from a VectoPostprocessor. I'm currently accessing the vector with the function 

```
 const VectorPostprocessorValue & c_var = getVectorPostprocessorValueByName(getParam<VectorPostprocessorName>("vpp"),getParam<std::string>("vector_name")); 
```

 but unfortunately I don't get the desired vector. Below, I attach my .C and .i.
The current script runs without any error, however the results I get are not correct. Do you have any suggestion please?  Thanks a lot for your help 

```
#include "NeckValue.h"
#include "NeckLineValueSampler.h"  
#include "SystemBase.h" 
using namespace std;   

registerMooseObject("ParrotApp", NeckValue);

InputParameters
NeckValue::validParams()
{
  InputParameters params = GeneralPostprocessor::validParams();
params.addRequiredParam<Real>("threshold", "Seconda variabile");
params.addRequiredParam<Real>("tolerance", "Terza variabile");
params.addRequiredParam<VectorPostprocessorName>("vpp", "The name of the VectorPostprocessor to pull the data from."); 
params.addRequiredParam<std::string>("vector_name","The name of the vector to use from the VectorPostprocessor"); 
params.addParam<std::vector<Real>>("c_var", std::vector<Real>{}, "Vector of real values used in computation.");
  return params;
}


NeckValue::NeckValue(
  const InputParameters &parameters)  
    : GeneralPostprocessor(parameters), 
    _threshold(getParam<Real>("threshold")),
    _tolerance(getParam<Real>("tolerance")),
    _c_var(getParam<std::vector<Real>>("c_var")) 
{}

void 
NeckValue::initialize()
{
 _c_var = getParam<std::vector<Real>>("c_var"); 
}
void 
NeckValue::execute()
{
  
}


Real NeckValue::computeValue() { 

  _value = 0.0; 
  _index = 0.0;
 
  const VectorPostprocessorValue & c_var = getVectorPostprocessorValueByName(getParam<VectorPostprocessorName>("vpp"),getParam<std::string>("vector_name"));
  
  for (unsigned int i = 0; i < _c_var.size(); i++)
  {
    if (std::abs(_c_var[i] - _threshold) < _tolerance)  
    {
      _value = c_var[i]; 
      _index = i; 
      break;  
    } 
  }
                       
    return _value;  
}  

  PostprocessorValue NeckValue::getValue() const       
  {                                         
    return _value;                              
  }
```



```
[VectorPostprocessors]
  [./c_along_vertical]
    type = NeckLineValueSampler
    variable = c                    
    start_point = '40 0 0'          
    end_point = '40 30 0'           
    num_points = 100                
    sort_by = y 
  [../] 
[]


[Postprocessors]
  [./c_neck]
    type = NeckValue
    vpp = 'c_along_vertical' 
    threshold = 0.4 
    tolerance = 0.05
  [../]  
[]
```

**Reply by GiudGiud** [reply]
Is it lagged?
Do you have a print out of a the vector in the VPP and in the PP?

**Reply by TFdvd** [reply]
In my VPP I get the output in a .csv like this: 
[variable] [id] [x] [y] [z]. 
In PP I would like to take the vector [variable] and manipulate it. My difficulty is to export [variable] from VPP to PP.

**Reply by GiudGiud** [reply]
What error do you get?

**Reply by TFdvd** [reply]
I don't get an error. The results are not coherent and I think that my procedure is wrong when I pass [variable] from VPP to PP. Maybe, passing the vector "[variable]" in the right way, I will solve the problem.

**Reply by GiudGiud** [reply]
Can you report here the print out of the vector in the VPP and in the PP?

**Reply by TFdvd** [reply]
Yes, sure!
A generic .csv of my VPP is:

[C_along_vertical_0005_.xlsx](https://github.com/user-attachments/files/17376258/C_along_vertical_0005_.xlsx)

The file above treats the i-th iteration. On the previous file, that comes from VPP, I'm only interested on the entire vector c (1st column) that I want to export on the PP in order to catch one only value of my interest.

The result of my PP for all iteration in the single value of my interest is:

[c_neck.xlsx](https://github.com/user-attachments/files/17376261/c_neck.xlsx)

As you can see, the repeted value 8.8931816251424e-323 in c_neck.xlsx file is not coherent and maybe VPP and PP are not comunicate well from the previous .C and .i that I attached. Thanks for your patience @GiudGiud

**Reply by GiudGiud** [reply]
```
 8.8931816251424e-323
```
means an unitialized value in C++

what is happening is that nothing ever calls computeValue().
You need to do this
```
void 
NeckValue::execute()
{
  computeValue();
}
```

or move the code in computeValue into execute

**Comment by TFdvd** [comment]
Hello @GiudGiud! 
I really want to thank you for your help, my custom PP works almost 100%. I have solved almost all the problems except for one detail that doesn't add up. From the results I attach below you can see that at the first iteration 0.2 I get 3.22 e-320 while at iteration 0.5 I get 4.42, which is the result I need to get in 0.2. In other words, I get the results of the current iteration at the next iteration.

Time  |    c
  0      |    0
  0.2   |    3.56e-321
  0.5   |    4.84
 0.95  |   6.06
   .      |      .
   .      |      .
   .      |      ...... and so on

I have tried indexing the for loop in various ways without getting a solution. I attach the .C below. Do you have any suggestions or notice anything that might explain this tendency? The explanation I have found is that perhaps the PP runs the loop before the VPP and so at the current iteration the PP finds no value from the VPP, while at the next iteration the PP finds data from the VPP which is still at the previous iteration. I don't know if this makes sense, but I would like to know your point of view and how I could solve this last detail. Thanks again for your help @GiudGiud ! 

 
#include "NeckValue.h"    

registerMooseObject("ParrotApp", NeckValue);

InputParameters
NeckValue::validParams()
{
  InputParameters params = GeneralPostprocessor::validParams();
  params.addRequiredParam<Real>("threshold", "Seconda variabile");
  params.addRequiredParam<Real>("tolerance", "Terza variabile");
  params.addRequiredParam<VectorPostprocessorName>("vpp", "The name of the VectorPostprocessor to pull the data from."); //new
  params.addRequiredParam<std::string>("vector_name","The name of the vector to use from the VectorPostprocessor"); //new
  params.addParam<std::vector<Real>>("c_var", std::vector<Real>{}, "Vector of real values used in computation.");
  return params;
}


NeckValue::NeckValue(
  const InputParameters &parameters)  
    : GeneralPostprocessor(parameters), 
    _threshold(getParam<Real>("threshold")),
    _tolerance(getParam<Real>("tolerance")),
    _c_var(getParam<std::vector<Real>>("c_var")) //virgola in più
    
{}

void 
NeckValue::initialize()
{

 _c_var = getParam<std::vector<Real>>("c_var"); 

}

void  
NeckValue::execute()
{
  computeValue(); 
}


Real NeckValue::computeValue() { 

  const VectorPostprocessorValue & _c_var = getVectorPostprocessorValueByName(getParam<VectorPostprocessorName>("vpp"),getParam<std::string>("vector_name"));
  
  for (unsigned int i = 0; i < _c_var.size(); i++)
  
  {
    if (std::abs(_c_var[i] - _threshold) < _tolerance)  
    {
      
      _value = _c_var[i]; 

[ ] 

      _index = i;

      _true_neck = (_index * 0.303); 

      break; 
    }  
  }

    return _true_neck; 
}

 
  PostprocessorValue NeckValue::getValue() const      
  {  
    return _true_neck;                         
                                                 
  }

**Reply by GiudGiud** [reply]
ok so what must be happening is that the postprocessor is executed before the VPP

so to fix that, use execution_order_groups (a parameter for both PPs, VPPs and UOs), and set the PP to be executed after the VPP

**Reply by TFdvd** [reply]
Perhaps I have not understood your point of view. 
The VPP has to be executed first in order to provide the data to the PP afterwards. 
If it is a matter of using execution_order_groups anyway, could you explain how to do that? Thanks in advance

**Reply by GiudGiud** [reply]
in the VPP set
execution_order_group = 0
in the PP set
execution_order_group = 1

that should work

**Reply by TFdvd** [reply]
You're right!! Thank a lot @GiudGiud
